Refactor wagtailadmin/shared/field.html into a template tag

gasman opened the pull request:
`wagtailadmin/shared/field.html` is pretty spaghetti-ish and contains lots of variable-shuffling logic that doesn't really belong in a template, in order to accommodate the variety of ways it can be invoked. Address this by converting it into a new template tag `{% formattedfield %}` which does that logic in Python code instead, with the added bonus of making form template code throughout Wagtail that little bit snappier.

This should remain backwards-compatible with all third-party code that uses `{% include "wagtailadmin/shared/field.html" %}`, as field.html now funnels its output through the new tag while preserving the fishing-around-in-template-context API. The only thing that this breaks (hopefully) is using `{% extends %}` to override the HTML of the actual field, but the tag-pair variant of the template tag can be used for this instead (was `{% field %}...{% endfield %}`, now `{% rawformattedfield %}...{% endrawformattedfield %}`).

Didn't get quite as far on this as I'd like - was hoping to fully deprecate use of wagtailadmin/shared/field.html, but FieldPanel and the rawformattedfield tag still rely on it. This is probably enough of an improvement (and a chunky enough PR) to be reviewing as it is, though...

I think the next step would be to move the variable-shuffling into a template component, so that `formattedfield` and `rawformattedfield` are effectively just wrappers that instantiate and immediately render the component (and can each have a trimmed-down signature for dealing with field objects and raw HTML respectively, rather than `formattedfield` being the kitchen sink), and FieldPanel can work with the component directly.

laymonage commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -1194,11 +1198,118 @@ class PanelNode(BlockInclusionNode):
+
+        errors = field.errors
+        has_errors = bool(errors)
+        if has_errors and hasattr(field.field.widget, "render_with_errors"):
```
Perhaps we should reuse the code for `has_unrendered_errors` here, to align with what the template originally had (and avoid the weirdness of `field.field` because `field` is technically a `BoundField`).

```suggestion
        if has_errors and not has_unrendered_errors(field):
```

gasman replied:
To be honest, the existence of `has_unrendered_errors` has always bugged me - a bit of one-off rendering logic to say "if this field has errors and doesn't handle rendering them for itself, render them here" that I would rather have placed directly in the template but couldn't, so had to come up with a tortuously-named filter to capture the idea. Being able to write that logic inline in actual Python code is scratching a very long-standing itch :-)

I can't remember why I stopped short of removing `has_unrendered_errors` completely here - maybe it was out of concern that third-party code might be using it. Either way, I'd be in favour of us deprecating it at the same time as deprecating direct use of field.html.

laymonage replied:
OK, if the plan is to remove `has_unrendered_errors` then I'm all for it, as that's what I'd prefer too.

laymonage commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -1290,6 +1288,32 @@ def formattedfield(
+    wagtailadmin/shared/field.html template.
+    """
+    kwargs = {}
+    for arg in (
```
A matter of preference but we could define the args in a list and use dict comprehension to make this more declarative :D

laymonage commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -1194,11 +1198,118 @@ class PanelNode(BlockInclusionNode):
+
+
+@register.inclusion_tag("wagtailadmin/shared/formatted_field.html", takes_context=True)
+def formattedfieldfromcontext(context):
```
The name gets quite long here, would something like `contextformattedfield` work better (to align with `rawformattedfield`)?

And I just want to point out that we have tags/filters with single "word" names, and we also have ones that use underscores. I don't really want to discuss which we should prefer right now though!

laymonage reviewed (commented):
I'm happy with the overall idea of the changes here. I have a few minor comments and I'll still have to test things out to make sure we don't break things, so this is not a final review yet. Thanks!

thibaudcolas commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -1194,11 +1198,118 @@ class PanelNode(BlockInclusionNode):
 
 
-register.tag("field", FieldNode.handle)
+register.tag("rawformattedfield", RawFormattedFieldNode.handle)
```
This is a bit hard to read as a single word all in lowercase. Could we add underscores?

```suggestion
register.tag("raw_formatted_field", RawFormattedFieldNode.handle)
```

gasman replied:
I thought that might be a little bit contentious... I went with the underscore-less version because it's consistent with how BlockInclusionNode generates the end tag by prepending `end` with no delimiter. As it stands, that would give us `endraw_formatted_field` rather than the expected `end_raw_formatted_field`. If we're changing that, we should probably update BlockInclusionNode to allow overriding that.

(Django also seems to have a slight preference for the no-underscore version in its own tags - e.g. `blocktranslate` - although it isn't really consistent.)

thibaudcolas commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -1194,11 +1198,118 @@ class PanelNode(BlockInclusionNode):
+
+
+@register.inclusion_tag("wagtailadmin/shared/formatted_field.html")
+def formattedfield(
```
Same comment as above re readability of the tag name. It’d also be nice if tag name and template name used the same case where possible.

thibaudcolas commented on wagtail/admin/templates/wagtailadmin/shared/formatted_field.html:
```diff
@@ -0,0 +1,61 @@
+
+    {# Render custom label attributes if provided, or the bound field’s label attributes otherwise. #}
+    {% if show_label %}
+        <label class="w-field__label{% if sr_only_label %} w-sr-only{% endif %}"{% if label_for %} for="{{ label_for }}"{% endif %}{% if label_id %} id="{{ label_id }}"{% endif %}>
```
```suggestion
        {# Add an id to the label so we can use it as a descriptor for the "Add comment" button. #}
        <label class="w-field__label{% if sr_only_label %} w-sr-only{% endif %}"{% if label_for %} for="{{ label_for }}"{% endif %}{% if label_id %} id="{{ label_id }}"{% endif %}>
```

This felt like an important comment to me, as it’s otherwise really unusual to have an id on a label element.

thibaudcolas reviewed (changes requested):
The refactoring of the tags is all looking good! I have some concerns about readability of the lowercasetagnames.

Outside the changes here:

- I see three uses of `field.html` which I’m not sure why we’re retaining
- We’ll need to update the comment in `FieldBlock.js` to refer to the correct template.

thibaudcolas reviewed (approved):
Just noting I’m happy with the rationale for the tag name

gasman commented on wagtail/admin/templates/wagtailadmin/shared/field.html:
```diff
@@ -1,88 +1,2 @@
-    </div>
-</div>
+{% load wagtailadmin_tags %}
+{% formattedfield field=field rendered_field=rendered_field classname=classname show_label=show_label id_for_label=id_for_label sr_only_label=sr_only_label icon=icon help_text=help_text help_text_id=help_text_id show_add_comment_button=show_add_comment_button label_text=label_text error_message_id=error_message_id %}
```
I'd rather avoid having an intermediate commit that temporarily breaks things. (I'm a little surprised that git doesn't automatically match the file content and handle it as a rename though...)

laymonage replied:
I think it's fine as long as it's still in the same PR, I doubt we really enforce a rule where "every single commit doesn't break". As someone who tends to do a bit of code archaeology when making changes to core things like this, it's so much easier to follow when git/GitHub detects that the file has been renamed.

But that's just me, I'm happy either way!

gasman commented:
Rebased and updated:
* Replaced two more references to field.html (wagtailadmin/collection_privacy/set_privacy.html and wagtailadmin/permissions/includes/collection_member_permissions_form.html) with formatted_field - not sure why I missed those before...
* Reinstated the comment re the label ID
* Updated the mention of field.html in FieldBlock.js

laymonage reviewed (approved):
Looks good to me! Although it looks like we'll have to update some docs i.e.

https://github.com/wagtail/wagtail/blob/456515d50049c50ecd8a5fb3484b41509e99e517/docs/advanced_topics/customisation/custom_user_models.md#L56-L57

https://github.com/wagtail/wagtail/blob/456515d50049c50ecd8a5fb3484b41509e99e517/docs/advanced_topics/customisation/custom_user_models.md#L67-L68

https://github.com/wagtail/wagtail/blob/456515d50049c50ecd8a5fb3484b41509e99e517/docs/extending/custom_account_settings.md?plain=1#L144

https://github.com/wagtail/wagtail/blob/456515d50049c50ecd8a5fb3484b41509e99e517/docs/extending/customising_group_views.md?plain=1#L85

gasman commented:
@laymonage That's part of the future plan that I didn't have chance to fully implement here - we can't really tell people to switch to the new tag until we have a deprecation process in place for the current template, and we can't do that until we've migrated FieldPanel away from it.
